This is dataset of used cars from eBay Kleinanzeigen, a classifieds 
section of the German eBay website.

The dataset was originally scraped and uploaded to Kaggle. 

There are sampled 50,000 data points from the full dataset, for code to
run quickly.
This is dirtied the dataset to more closely resemble what you would expect
from a scraped dataset.


In [256]:
import pandas as pd
import numpy as np
autos = pd.read_csv('autos.csv', encoding = "Latin-1")

In [257]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [258]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [259]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [260]:
#editing column names
autos.rename( {
    'yearOfRegistration':'registration_year',
    'monthOfRegistration':'registration_month',
    'notRepairedDamage':'unrepaired_damage',
    'dateCreated':'ad_created',
    'dateCrawled':'date_crawled'
    }, axis = 1, inplace = True )
#lowering all columns
new_columns = []
for column in autos.columns:
    lowered_columns = column.lower()
    new_columns.append(lowered_columns)
autos.columns = new_columns
autos.head()
autos.rename( {
    'lastseen':'last_seen',
    'postalcode':'postal_code',
    'fueltype': 'fuel_type',
    'offertype': 'offer_type',
    'vehicletype':'vehicle_type',
    'abtest': 'abtest'
    }, axis = 1, inplace = True )


I changed column names to more standard and readeble name. Also I changed column names from camelcase to snakecase to never worry about format of letter

Now let's do some basic data exploration to determine what other cleaning 
tasks need to be done. Initially we will look for: - Text columns where 
all or almost all values are the same. These can often be dropped as they 
don't have useful information for analysis. - Examples of numeric data 
stored as text which can be cleaned and converted.

In [261]:
autos.describe()

registration_year       powerps  registration_month  nrofpictures  \
count       50000.000000  50000.000000        50000.000000       50000.0   
mean         2005.073280    116.355920            5.723360           0.0   
std           105.712813    209.216627            3.711984           0.0   
min          1000.000000      0.000000            0.000000           0.0   
25%          1999.000000     70.000000            3.000000           0.0   
50%          2003.000000    105.000000            6.000000           0.0   
75%          2008.000000    150.000000            9.000000           0.0   
max          9999.000000  17700.000000           12.000000           0.0   

        postal_code  
count  50000.000000  
mean   50813.627300  
std    25779.747957  
min     1067.000000  
25%    30451.000000  
50%    49577.000000  
75%    71540.000000  
max    99998.000000

Most likely, we shall drop nrofpictures (basicaly no values there) and datecrawled (equal to ad_created) columns. We shall need to make new digit columns, remove "$" symbol from price, "km" from odometer column

In [262]:
#removing '$' from price and stripping of extra spaces
autos['price'] = (autos['price']
                  .str.replace("$","")
                  .str.replace(",","")
                  .str.strip()
                 )
#removing 'km' from odometer and stripping of extra spaces
autos['odometer'] = (autos['odometer']
                     .str.replace("km","")
                     .str.replace(",","")
                     .str.strip()
                    )
#renaming columns
autos.rename( {'odometer':'odometer_km'}, axis = 1, inplace = True)
#making both columns numeric type
autos['price'] = autos['price'].astype(int)
autos['odometer_km'] = autos['odometer_km'].astype(int)


In [263]:
autos.describe()

price  registration_year       powerps    odometer_km  \
count  5.000000e+04       50000.000000  50000.000000   50000.000000   
mean   9.840044e+03        2005.073280    116.355920  125732.700000   
std    4.811044e+05         105.712813    209.216627   40042.211706   
min    0.000000e+00        1000.000000      0.000000    5000.000000   
25%    1.100000e+03        1999.000000     70.000000  125000.000000   
50%    2.950000e+03        2003.000000    105.000000  150000.000000   
75%    7.200000e+03        2008.000000    150.000000  150000.000000   
max    1.000000e+08        9999.000000  17700.000000  150000.000000   

       registration_month  nrofpictures   postal_code  
count        50000.000000       50000.0  50000.000000  
mean             5.723360           0.0  50813.627300  
std              3.711984           0.0  25779.747957  
min              0.000000           0.0   1067.000000  
25%              3.000000           0.0  30451.000000  
50%              6.000000           0.0  49577.000000  
75%              9.000000           0.0  71540.000000  
max             12.000000           0.0  99998.000000

In [264]:
def description(series, tf = False):
    #this function gives overall description over digit columns
    print('unique values', series, "\n", autos[series].unique().shape, "\n")
    print(series, 'parameters', "\n", autos[series].describe(), "\n")
    print('value counts', series, "\n", autos[series].
      value_counts(dropna=False)
      .head()
      .sort_index(ascending = tf))
    return series

description('odometer_km', False)
description('price', True)   
autos["price"].value_counts().sort_index(ascending=False).head(20)


unique values odometer_km 
 (13,) 

odometer_km parameters 
 count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64 

value counts odometer_km 
 150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
Name: odometer_km, dtype: int64
unique values price 
 (2357,) 

price parameters 
 count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64 

value counts price 
 0       1421
500      781
1000     639
1500     734
2500     643
Name: price, dtype: int64


99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64

In [265]:
#making filter over data
def filter(column, x, y):
    print(autos[autos[column].between(x,y)])
    return column

filter('odometer_km', 100000,110000)
filter('price', 10000,20000)
    

              date_crawled                                               name  \
68     2016-04-03 17:36:59  Mini_cooper_s_clubman_/vollausstattung_/_Navi/...   
125    2016-03-17 17:56:46                                     Seat_Altea_1_6   
131    2016-04-03 14:49:09                                  Audi_A3_Sportback   
149    2016-03-12 13:51:26                                     Kia_Sephia_GTX   
160    2016-03-31 02:36:24                                         BMW_M5_DKG   
170    2016-03-17 14:38:23    Fiat_Seicento_1.1_Top_zustand_mit_tuev_servo___   
184    2016-04-04 23:56:50  Peugeot_207_75_Urban_Move__Panoramadach__Klima...   
219    2016-03-09 14:45:23                                    Opel_Zafira_2.2   
222    2016-03-27 15:58:44                                  Opel_Astra_Cabrio   
249    2016-03-18 17:37:34                   Dodge_Caliber_SXT_1.8_Benzin_Gas   
271    2016-03-20 18:42:22                          Audi_A4_Avant_2.0_TDI_DPF   
275    2016-03-07 18:41:56  

'price'

To include missing values in the distribution and to use percentages instead of counts, we chain the Series.value_counts(normalize=True, dropna=False) method.
To rank by date in ascending order (earliest to latest), we chain the Series.sort_index() method.

In [266]:
def normalizing_function(column_name):
    print(autos[column_name]
            .str[:10]
            .value_counts(normalize=True, dropna=False)
            .sort_index(ascending=False)
            )
    return column_name
normalizing_function('date_crawled')
normalizing_function('last_seen')


normalizing_function('ad_created')





2016-04-07    0.00142
2016-04-06    0.00318
2016-04-05    0.01310
2016-04-04    0.03652
2016-04-03    0.03868
2016-04-02    0.03540
2016-04-01    0.03380
2016-03-31    0.03192
2016-03-30    0.03362
2016-03-29    0.03418
2016-03-28    0.03484
2016-03-27    0.03104
2016-03-26    0.03248
2016-03-25    0.03174
2016-03-24    0.02910
2016-03-23    0.03238
2016-03-22    0.03294
2016-03-21    0.03752
2016-03-20    0.03782
2016-03-19    0.03490
2016-03-18    0.01306
2016-03-17    0.03152
2016-03-16    0.02950
2016-03-15    0.03398
2016-03-14    0.03662
2016-03-13    0.01556
2016-03-12    0.03678
2016-03-11    0.03248
2016-03-10    0.03212
2016-03-09    0.03322
2016-03-08    0.03330
2016-03-07    0.03596
2016-03-06    0.01394
2016-03-05    0.02538
Name: date_crawled, dtype: float64
2016-04-07    0.13092
2016-04-06    0.22100
2016-04-05    0.12428
2016-04-04    0.02462
2016-04-03    0.02536
2016-04-02    0.02490
2016-04-01    0.02310
2016-03-31    0.02384
2016-03-30    0.02484
2016-03-29    0.022

'ad_created'

In [267]:
print(autos["ad_created"].str[:10].unique().shape)

(76,)


In [268]:
autos['registration_year'].describe()

count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

In [269]:
autos["price"].value_counts().sort_index(ascending=True).head(20)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price, dtype: int64

In [270]:
autos = autos[autos["price"].between(1,351000)]
autos["price"].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

There is a large variety of ad created dates. Most fall within 1-2 months of the listing date, but a few are quite old, with the oldest at around 9 months.



We use Series.describe() to understand the distribution of registration_year.

In [271]:
autos['registration_month'].describe()

count    48565.000000
mean         5.782251
std          3.685595
min          0.000000
25%          3.000000
50%          6.000000
75%          9.000000
max         12.000000
Name: registration_month, dtype: float64

In [272]:
autos['registration_year'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The year that the car was first registered will likely indicate the age of the car. Looking at this column, we note some odd values. The minimum value is 1000, long before cars were invented and the maximum is 9999, many years into the future.

he highest and lowest acceptable values will be defined as 1950 and 2019 for the registration_year column. It is unlikely that there is still a car that has registration more than 60 years old. Even if there is, there must be small ammount of that cars and they should not effect our observations.

In [273]:
(~autos["registration_year"].between(1900,2020)).sum() / autos.shape[0]

0.00041181921136621024

In [274]:
autos = autos[autos['registration_year'].between(1900,2020)]


In [275]:
autos['registration_year'].value_counts(normalize=True)

2000    0.065012
2005    0.060480
1999    0.059677
2004    0.055680
2003    0.055598
2006    0.055001
2001    0.054300
2002    0.051210
1998    0.048676
2007    0.046905
2008    0.045628
2009    0.042950
1997    0.040190
2011    0.033433
2010    0.032733
2017    0.028674
1996    0.028283
2012    0.026985
1995    0.025276
2016    0.025131
2013    0.016541
2014    0.013657
1994    0.012957
2018    0.009682
1993    0.008755
2015    0.008075
1992    0.007622
1990    0.007148
1991    0.006983
1989    0.003584
          ...   
1976    0.000433
1969    0.000391
1975    0.000371
1965    0.000350
1964    0.000247
1963    0.000165
1959    0.000124
1961    0.000124
1910    0.000103
1937    0.000082
1962    0.000082
1958    0.000082
1956    0.000082
1950    0.000062
1955    0.000041
2019    0.000041
1934    0.000041
1957    0.000041
1951    0.000041
1941    0.000041
1954    0.000041
1953    0.000021
1948    0.000021
1927    0.000021
1943    0.000021
1929    0.000021
1939    0.000021
1938    0.0000

Seems like most cars were registed in 2000, as expected, the older the car the less percentage it has in registry

the next step in our research is data aggregation.
We shall:
- Identify the unique values we want to aggregate by
- Create an empty dictionary to store our aggregate data
- Loop over the unique values, and for each:
    - Subset the dataframe by the unique values
    - Calculate the mean of whichever column we're interested in
    - Assign the val/mean to the dict as k/v.

In [276]:
autos['brand'].value_counts(normalize = True).head(20)

volkswagen        0.212813
opel              0.108641
bmw               0.108641
mercedes_benz     0.095787
audi              0.085858
ford              0.069667
renault           0.047894
peugeot           0.029457
fiat              0.025996
seat              0.018931
skoda             0.016068
nissan            0.015264
mazda             0.015223
smart             0.014296
citroen           0.014111
toyota            0.012586
hyundai           0.009950
sonstige_autos    0.009641
volvo             0.009043
mini              0.008611
Name: brand, dtype: float64

I'm aggregating for top 20 brends in dataset

In [277]:
mean_price_by_brand = {}


In [278]:
brands = autos['brand'].value_counts(normalize = True).head(20).index
brands

Index(['volkswagen', 'opel', 'bmw', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'nissan', 'mazda', 'smart',
       'citroen', 'toyota', 'hyundai', 'sonstige_autos', 'volvo', 'mini'],
      dtype='object')

In [279]:
for brand in brands:
    selected_brand = autos[autos['brand'] == brand]
    mean_price = selected_brand['price'].mean()
    mean_price_by_brand[brand] = int(mean_price)
mean_price_by_brand

{'audi': 9212,
 'bmw': 8261,
 'citroen': 3756,
 'fiat': 2793,
 'ford': 3728,
 'hyundai': 5371,
 'mazda': 4059,
 'mercedes_benz': 8526,
 'mini': 10541,
 'nissan': 4669,
 'opel': 2941,
 'peugeot': 3065,
 'renault': 2431,
 'seat': 4320,
 'skoda': 6353,
 'smart': 3518,
 'sonstige_autos': 12208,
 'toyota': 5148,
 'volkswagen': 5333,
 'volvo': 4866}

most expencive brand sonstige_autos, least expensive - fiat

In [280]:
top_6_brands = autos['brand'].value_counts(normalize = True).head(6).index
mean_price_d = {}
mean_mileage = {}
for brand in top_6_brands:
    selected_brand = autos[autos['brand'] == brand]
    mean_price = selected_brand['price'].mean()
    mean_price_d[brand] = int(mean_price)
for brand in top_6_brands:
    selected_brand = autos[autos['brand'] == brand]
    mean_odometer = selected_brand['odometer_km'].mean()
    mean_mileage[brand] = int(mean_odometer)


In [281]:
top_6_brands = pd.Series(top_6_brands)

In [286]:
df = pd.DataFrame(top_6_brands,columns =['mean_price_d'] )
df["mean_mileage"] = mean_mileage 

In [287]:
df

mean_price_d   mean_mileage
0     volkswagen     volkswagen
1           opel           audi
2            bmw            bmw
3  mercedes_benz           ford
4           audi  mercedes_benz
5           ford           opel